### PR TITLE
Processor obtains documents from a Pub/Sub queue - NATS

### PR DESCRIPTION
**Pull request by pxp928**
Signed-off-by: pxp928 parth.psu@gmail.com

Added NATS Jetstream implementation to collect, process and ingest the documents. Pull based consumers are implemented that allow for scaling and allow multiple processors and ingestors to run simultaneously.

Need to setup e2e integration test for test NATS end to end.

To test locally using the example collector:
1. install nats-server: `brew install nats-server`
2. create js.conf file with the following:
```
// change max payload from default 1MB to 8MB
max_payload: 8MB
// enables jetstream, an empty block will enable and use defaults
jetstream {
}
```
3. Run `nats-server --config js.conf`
4. `make build` in the GUAC repo
5. Run ` ./bin/pubsub files --creds neo4j:s3cr3t <FILE PATH>`

Closes: #31

### CODE REVIEW

**Review by lumjjb — COMMENTED** [review]
flushing old comments

File: pkg/emitter/nats_emitter.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+// NATS stream
+const (
+	NatsName                string = "GUAC"
+	StreamName              string = "DOCUMENTS"
+	StreamSubjects          string = "DOCUMENTS.*"
+	SubjectNameDocCollected string = "DOCUMENTS.collected"
+	SubjectNameDocProcessed string = "DOCUMENTS.processed"
+	SubjectNameDocParsed    string = "DOCUMENTS.parsed"
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter.go:35]: nit: ingested for consistency?

File: pkg/emitter/nats_emitter.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+var (
+	nc *nats.Conn
+	js nats.JetStreamContext
+)
+
+func JetStreamInit(ctx context.Context, url string, creds string, nkeyFile string, insecure bool) nats.JetStreamContext {
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter.go:43]: nit: can have a config struct and pass it in, and put comments in the struct on what fields mean

File: pkg/emitter/nats_emitter.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	logger := logging.FromContext(ctx)
+	// Connect to NATS
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter.go:45]: nit: old duplicate comment?

File: pkg/emitter/nats_emitter.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	var err error
+	// Connect Options.
+	opts := []nats.Option{nats.Name(NatsName)}
+
+	// TODO: secure connection via User creds file or NKey file
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter.go:50]: is this still a TODO or is it done?

File: pkg/emitter/nats_emitter.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	}
+	err = createStream(ctx)
+	if err != nil {
+		logger.Fatalf("failed to create stream: %v", err)
+	}
+	return js
+
+}
+
+func createStream(ctx context.Context) error {
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter.go:85]: nit: createStreamOrExists

File: pkg/emitter/nats_emitter.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	logger := logging.FromContext(ctx)
+	stream, err := js.StreamInfo(StreamName)
+	if err != nil && !strings.Contains(err.Error(), "nats: stream not found") {
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter.go:88]: use `errors.Is` with `ErrStreamNotFound`

File: pkg/emitter/nats_emitter.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		}
+	}
+	return nil
+}
+
+func Emit(ctx context.Context, d *processor.Document) {
+	logger := logging.FromContext(ctx)
+	docByte, err := json.Marshal(d)
+	if err != nil {
+		logger.Warnf("failed marshal of document: %s", err)
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter.go:110]: should return in err case

File: pkg/emitter/nats_emitter.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	}
+	_, err = js.Publish(SubjectNameDocCollected, docByte)
+	if err != nil {
+		logger.Errorf("failed to publish document on stream: %v", err)
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter.go:114]: should return in err case

File: pkg/emitter/nats_emitter_test.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func TestNatsEmitter_PublishOnEmit(t *testing.T) {
+	s := RunServerOnPort(TEST_PORT)
+	err := s.EnableJetStream(&server.JetStreamConfig{})
+	if err != nil {
+		t.Fatalf("unexpected error initializing test NATS: %v", err)
+	}
+	defer s.Shutdown()
+	time.Sleep(time.Second * 5)
```
Comment (lumjjb) [inline pkg/emitter/nats_emitter_test.go:84]: annotate the sleep reason

File: pkg/handler/processor/process/process.go [hunk 10]
```diff
@@ -40,12 +45,57 @@ func RegisterDocumentProcessor(p processor.DocumentProcessor, d processor.Docume
...
+		return err
+	}
+	for {
+		// if the context is canceled we want to break out of the loop
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		msgs, err := sub.Fetch(1)
+		if err != nil {
+			logrus.Printf("[processor: %s] error consuming, sleeping for a second: %v", id, err)
```
Comment (lumjjb) [inline pkg/handler/processor/process/process.go:62]: nit: s/sleeping/backing off/

File: pkg/handler/processor/process/process.go [hunk 11]
```diff
@@ -40,12 +45,57 @@ func RegisterDocumentProcessor(p processor.DocumentProcessor, d processor.Docume
 	return nil
 }
```
Comment (lumjjb) [inline pkg/handler/processor/process/process.go:47]: can probably be in its own package?

File: pkg/emitter/nats_emitter.go (hunk 1 above)
Comment (pxp928) [reply pkg/emitter/nats_emitter.go:35]: removed as not being used.

File: pkg/emitter/nats_emitter.go (hunk 2 above)
Comment (pxp928) [reply pkg/emitter/nats_emitter.go:43]: done.

File: pkg/emitter/nats_emitter.go (hunk 3 above)
Comment (pxp928) [reply pkg/emitter/nats_emitter.go:45]: removed

File: pkg/emitter/nats_emitter.go (hunk 4 above)
Comment (pxp928) [reply pkg/emitter/nats_emitter.go:50]: removed as done.

File: pkg/emitter/nats_emitter.go (hunk 6 above)
Comment (pxp928) [reply pkg/emitter/nats_emitter.go:88]: done.

File: pkg/emitter/nats_emitter.go (hunk 7 above)
Comment (pxp928) [reply pkg/emitter/nats_emitter.go:110]: done.

File: pkg/emitter/nats_emitter.go (hunk 8 above)
Comment (pxp928) [reply pkg/emitter/nats_emitter.go:114]: done.

File: pkg/handler/processor/process/process.go (hunk 11 above)
Comment (pxp928) [reply pkg/handler/processor/process/process.go:47]: breakout the `Subscribe` to be its own package?

**Comment by pxp928** [comment]
Closing PR as this will be broken out into smaller PRs to allow for easier review and merge.

**Pull request closed without merging**